lib/src/front_end/ast_node_visitor.dart

```diff
@@ -927,17 +930,87 @@ class AstNodeVisitor extends ThrowingAstVisitor<void>
 
   @override
   void visitSwitchExpression(SwitchExpression node) {
-    throw UnimplementedError();
+    var list = DelimitedListBuilder.switchBody(this);
+
+    createSwitchValue(node.switchKeyword, node.leftParenthesis, node.expression,
+        node.rightParenthesis);
+    writer.space();
+    list.leftBracket(node.leftBracket);
+
+    for (var member in node.cases) {
+      list.add(member);
+    }
+
+    list.rightBracket(node.rightBracket);
+    writer.push(list.build());
   }
 
   @override
   void visitSwitchExpressionCase(SwitchExpressionCase node) {
-    throw UnimplementedError();
+    if (node.guardedPattern.whenClause != null) throw UnimplementedError();
+
+    visit(node.guardedPattern.pattern);
+    writer.space();
+    finishAssignment(node.arrow, node.expression);
   }
 
   @override
   void visitSwitchStatement(SwitchStatement node) {
-    throw UnimplementedError();
+    createSwitchValue(node.switchKeyword, node.leftParenthesis, node.expression,
+        node.rightParenthesis);
+
+    // Attach the ` {` as part of the `)` in the switch value ListPiece.
```

Comment:
I don't understand the mention of  `ListPiece`. Can you expand this?

Reply:
Done.